dag-cbor/src/encode.rs

```diff
@@ -171,109 +168,118 @@ impl WriteCbor for f64 {
     }
 }
 
-impl WriteCbor for [u8] {
+impl Encode<DagCbor> for [u8] {
     #[inline]
-    fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
         write_u64(w, 2, self.len() as u64)?;
         w.write_all(self)?;
         Ok(())
     }
 }
 
-impl WriteCbor for str {
+impl Encode<DagCbor> for Box<[u8]> {
+    #[inline]
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
+        Encode::<DagCbor>::encode(&self[..], w)
+    }
+}
+
+impl Encode<DagCbor> for str {
     #[inline]
-    fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
         write_u64(w, 3, self.len() as u64)?;
         w.write_all(self.as_bytes())?;
         Ok(())
     }
 }
 
-impl WriteCbor for String {
+impl Encode<DagCbor> for String {
     #[inline]
-    fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
-        self.as_str().write_cbor(w)
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
+        self.as_str().encode(w)
     }
 }
 
-impl WriteCbor for i128 {
+impl Encode<DagCbor> for i128 {
     #[inline]
-    fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
         if *self < 0 {
             if -(*self + 1) > u64::max_value() as i128 {
-                return Err(CborError::NumberOutOfRange);
+                return Err(Error::NumberOutOfRange);
             }
             write_u64(w, 1, -(*self + 1) as u64)?;
         } else {
             if *self > u64::max_value() as i128 {
-                return Err(CborError::NumberOutOfRange);
+                return Err(Error::NumberOutOfRange);
             }
             write_u64(w, 0, *self as u64)?;
         }
         Ok(())
     }
 }
 
-impl WriteCbor for Cid {
+impl Encode<DagCbor> for Cid {
     #[inline]
-    fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
+    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
         write_tag(w, 42)?;
         // insert zero byte per https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md#links
-        let bytes = self.to_bytes();
-        write_u64(w, 2, (bytes.len() + 1) as u64)?;
+        // 128 bytes should be plenty large enough for any cid
+        let mut buf = [0u8; 128];
```

Comment:
There is also the identity hash, which is just the data itself. There 128 byte wouldn't be enough.

Reply:
the identity hash doesn't have a size restriction?

Reply:
I believe that would be a bug. The reason for the identity hash is when the data is smaller than a cryptographic hash function. So it can not be longer than the longest hash function we have which would be 512bits.

https://github.com/multiformats/multihash/issues/13